controllers/kustomization_controller.go

```diff
@@ -615,47 +621,90 @@ func (r *KustomizationReconciler) prune(kustomization kustomizev1.Kustomization,
 	return nil
 }
 
+func toObjMetadata(wr []kustomizev1.WorkloadReference) []object.ObjMetadata {
+	oo := []object.ObjMetadata{}
+	for _, w := range wr {
+		oo = append(oo, object.ObjMetadata{
+			Name:      w.Name,
+			Namespace: w.Namespace,
+			GroupKind: schema.GroupKind{
+				Group: w.GroupKind.Group,
+				Kind:  w.GroupKind.Kind,
+			},
+		})
+	}
+	return oo
+}
+
+func toHealthySet(wr []kustomizev1.WorkloadReference) map[string]bool {
+	hs := map[string]bool{}
+	for _, w := range wr {
+		hs[w.String()] = false
+	}
+	return hs
+}
+
+func filterHealthSet(hs map[string]bool, healthy bool) []string {
+	res := []string{}
+	for k, v := range hs {
+		if v == healthy {
+			res = append(res, k)
+		}
+	}
+	return res
+}
+
 func (r *KustomizationReconciler) checkHealth(kustomization kustomizev1.Kustomization, revision string) error {
+	if len(kustomization.Spec.HealthChecks) == 0 {
+		return nil
+	}
+
 	timeout := kustomization.GetTimeout() + (time.Second * 1)
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
-	var alerts string
 
-	for _, check := range kustomization.Spec.HealthChecks {
-		cmd := fmt.Sprintf("until kubectl -n %s get %s %s ; do sleep 2; done",
-			check.Namespace, check.Kind, check.Name)
-		command := exec.CommandContext(ctx, "/bin/sh", "-c", cmd)
-		if _, err := command.CombinedOutput(); err != nil {
-			return fmt.Errorf("health check timeout for %s '%s/%s': %w",
-				check.Kind, check.Namespace, check.Name, err)
-		}
+	opts := polling.Options{PollInterval: 500 * time.Millisecond, UseCache: true}
+	healthySet := toHealthySet(kustomization.Spec.HealthChecks)
+	eventsChan := r.Poller.Poll(ctx, toObjMetadata(kustomization.Spec.HealthChecks), opts)
+
+	for {
+		select {
+		case <-ctx.Done():
+			notHealthy := filterHealthSet(healthySet, false)
+			return fmt.Errorf("Health check timeout for [%v]", strings.Join(notHealthy, ", "))
+		case e := <-eventsChan:
+			switch e.EventType {
+			case event.ResourceUpdateEvent:
+				id := fmt.Sprintf("%s/%s/%s", e.Resource.Identifier.GroupKind.String(), e.Resource.Identifier.Namespace, e.Resource.Identifier.Name)
+				if e.Resource.Status == status.CurrentStatus {
+					healthySet[id] = true
+					r.Log.WithValues(
+						strings.ToLower(kustomization.Kind),
+						fmt.Sprintf("%s/%s", kustomization.GetNamespace(), kustomization.GetName()),
+					).Info(fmt.Sprintf("Health check passed for %s", id))
+				} else {
+					healthySet[id] = false
+				}
+			case event.ErrorEvent:
+				return e.Error
+				// Event does not behave like expected and will not occur when all the
+				// resources are in a current state.
+				/*case event.CompletedEvent:
+				if alerts != "" && kustomization.Status.LastAppliedRevision != revision {
+					r.event(kustomization, revision, recorder.EventSeverityInfo, alerts)
+				}
+				return nil*/
```

Comment:
This issue needs to be fixed before the PR can be merged.

I have created an issue for this in  the cli-utils repo.
https://github.com/kubernetes-sigs/cli-utils/issues/232